### PR TITLE
Fix memory leak in ProjectCacheService...

### PR DESCRIPTION
**Pull request by KevinH-MS**
We create a SimpleMRUCache for all Workspace types, but we only monitor it
(clean up) for the Visual Studio Workspace.  If you repeatedly
opened/closed files in the Miscellaneous Files Workspace, we'd cache every
Solution and never clean them up (as described in DevDiv 126519).

Alternatively, we could consider only creating a cache in the ```WorkspaceKind.Host``` case.

### CODE REVIEW

**Comment by KevinH-MS** [comment]
FYI @heejaechang 
/cc @dotnet/roslyn-ide

**Comment by heejaechang** [comment]
simpleMRUcache only cache last 3 items. so not sure whether it is actually leak and not sure we need this cache for anything else than host, misc, metadata as source workspace. I am not sure what is right approach.

**Comment by jasonmalinowski** [comment]
What does this cache even cache?

**Comment by heejaechang** [comment]
well. long story. Roslyn used to have implicit cache only. and then Kalye suggested explicit cache is better for our perf number so, we removed all implicit cache and completely moved to explicit cache and that made perf number better for our perf tests but issue came up in other teams perf number so we decide put simple implicit cache and this is that cache.

I think @rchande implemented this simple cache so adding him in case he has something to add.

**Comment by heejaechang** [comment]
ah, I might misunderstand question. it is used for text, tree, compilation. we used to have 3 different cache for each, and when we moved to explicit cache, we merged all into 1. so I believe mostly for compilation, but caching compilation indirectly hold others now, so ..

**Comment by heejaechang** [comment]
it probably make some temporary workspace such as preview workspace to have that monitor created unnecessary but probably won't matter much.

**Comment by rchande** [comment]
The explicit cache only caches the compilation/trees for one project, and we have some heuristics for setting the active project. When we made that change, there was a big winforms regression, IIRC. We noticed then that the regression consisted of rebuilding compilations, we wound up with 3 because that fixed the regression without making the cache much bigger.

File: src/EditorFeatures/Core/Implementation/Workspaces/ProjectCacheService.cs [hunk 1]
```diff
@@ -17,19 +17,15 @@ internal partial class ProjectCacheService : IProjectCacheHostService
...
             _workspace = workspace;
 
-            // forceCleanup is for testing
-            if (workspace?.Kind == WorkspaceKind.Host || forceCleanup)
-            {
-                // monitor implicit cache for host
-                _implicitCacheMonitor = new ImplicitCacheMonitor(this, implicitCacheTimeout);
-            }
+            _implicitCache = new SimpleMRUCache();
+            _implicitCacheMonitor = new ImplicitCacheMonitor(this, implicitCacheTimeout);
```
Comment (rchande) [inline src/EditorFeatures/Core/Implementation/Workspaces/ProjectCacheService.cs]: Should we add a test for this?

**Comment by rchande** [comment]
If we only attach a cache to the VS workspace, is there any way for us to determine if Misc Files perf gets worse?

File: src/EditorFeatures/Core/Implementation/Workspaces/ProjectCacheService.cs (hunk 1 above)
Comment (KevinH-MS) [inline src/EditorFeatures/Core/Implementation/Workspaces/ProjectCacheService.cs]: Now that the ```if``` has been removed, we have a test that covers all code paths (```TestImplicitCacheMonitoring```).  If we add the ```if``` back, then yeah, we should have a test for other types of Workspaces.

**Comment by KevinH-MS** [comment]
@rchande I can't say that Misc Files perf won't get worse in some cases by removing the cache, but I can imagine the effects of the (potentially hundreds of megabytes) of data that we keep alive after opening several large misc files would outweigh the costs of reparsing (unless there's a scenario where you rapidly open/close the same few files).

It's sounding to me like the approach we'd rather take is to restrict the implicit cache to cases we know matter (```WorkspaceKind.Host```).  If that sounds right to you/HeeJae, I'll update the PR.

**Comment by heejaechang** [comment]
@KevinH-MS I like that approach better actually.

**Comment by rchande** [comment]
Yeah, that sounds good to me.

**Comment by KevinH-MS** [comment]
PR updated to only create ```SimpleMRUCache``` in the ```WorkspaceKind.Host``` case...

**Comment by heejaechang** [comment]
:+1:

**Comment by jasonmalinowski** [comment]
I wouldn't be worried about Miscellaneous files, since those projects are always pretty "small" -- each file is always in it's own project, and the files are always open, so things should be held around for other reasons.

Doesn't have to be done as a part of this bug, but can some doc comment be added onto these cache types with @heejaechang's description of what this caches? I don't mind that we have a cache or it's implementation thereof, my concern as simply that I have no clue what this actually holds onto looking at the code. "ProjectCache" makes me think it caches...Projects? But how?

**Comment by KevinH-MS** [comment]
Sorry, I realized I wasn't running all the tests I thought I was locally.  I had to fix up a couple more tests and add a couple new null checks.  @heejaechang @rchande if you could please review the second commit...

**Comment by KevinH-MS** [comment]
@jasonmalinowski in the case detailed in the bug, the cost of the Compilation associated with the Misc file was actually pretty substantial (tens of MB that added up to a couple hundred MB when you combined all 3 Compilations we were caching).  I will add a doc comment per your suggestion.

**Comment by heejaechang** [comment]
:+1:

**Comment by KevinH-MS** [comment]
test eta please

**Pull request merged**